### PR TITLE
Add new reco::Muon 'displacedMuons' collection to AOD

### PR DESCRIPTION
**Pull request by CeliaFernandez**
#### PR description:

This PR adds a new reco::Muon collection, called 'displacedMuons', that includes the dedicated displaced muon reconstructions: 'displacedTracks', 'displacedGlobalMuons' and 'displacedStandAloneMuons' (all of them saved in AOD as reco::Track collections).

In this PR we add the necessary modifications to:
1. Cross-clean the existing reco::Track objects in order to avoid the duplication or loss of displaced muon candidates in the output collection.
2. Compute the main muon variables, including isolation, time, recoTracksExtras... etc
3. Save this information in AOD and RECO, as configured for standard 'muons' collection

The sequences are implemented to run independently to the default reco::Muon 'muons' collection and avoid interferences. Muons of the 'muons' collection should remain untouchable. They also emulate the existing implementation and make use of the existing tools and sequences so the new information is computed with the default parameters configuration.

The following packages are modified:

- RecoMuon/Configuration: To perform the cross-cleaning (in RecoMuonPPOnly), integrate the new displaced sequences (RecoMuon_cff.py) and define the information saved in AOD (RecoMuon_EventContent_cff.py)
- RecoMuon/Identification: To define the main 'displacedMuons' sequence for the new collection (displacedMuons_cfi.py) and the sequence to compute the ReducedTrackExtras associated to it (displacedMuonsReducedTrackExtras_cfi.py)
- RecoMuon/MuonIsolationProducers: To add the detector-based isolation deposits (muIsolation_cff.py, muIsoDeposits_cff.py, muIsoDepositCopies_cfi.py and muIsolation_EventContent_cff.py)
- RecoMuon/MuonIsolation: To add the PFIsolation (muonPFIsolation_cff.py, displacedMuonPFIsolationValues_cff.py and displacedMuonPFIsolationDeposits_cff.py)

The increase in size was measured by using a ttbar benchmark model (runTheMatrix.py workflow 11634.0), where we obtained a relative increase of the 1.23%

Backup documentation, detailed information about the modifications and validation can be found here:  https://cernbox.cern.ch/index.php/s/FfAZonfSest7kqQ

The implementation was built on top of CMSSW_12_3_0_pre2

@gkaratha @sscruz @trocino 

#### PR validation:

We have checked out all dependencies and have a clean build.
The actions defined in the new sequences were validated and these results are presented in the backup documentation:
https://cernbox.cern.ch/index.php/s/FfAZonfSest7kqQ

### CODE REVIEW

File: RecoMuon/Configuration/python/RecoMuon_EventContent_cff.py [hunk 1]
```diff
@@ -36,6 +36,15 @@
                                            #'drop *_muons_muons1stStep2muonsMap_*',
                                            #'drop recoIsoDepositedmValueMap_muons_*_*', #not really used
                                            #'drop doubleedmValueMap_muons_muPFIso*_*', #already inside the muon
+                                           # displacedMuons collection
+                                           'keep recoMuons_displacedMuons_*_*',
```
Comment (clacaputo) [inline RecoMuon/Configuration/python/RecoMuon_EventContent_cff.py:40]: Do you know what is the contribution to the total size increase given by each of these new collections?

**Comment by clacaputo** [comment]
@cmsbuild please test

File: RecoMuon/Configuration/python/RecoMuon_EventContent_cff.py (hunk 1 above)
Comment (CeliaFernandez) [reply RecoMuon/Configuration/python/RecoMuon_EventContent_cff.py:40]: We have computed it in a ttbar sample (runTheMatrix.py workflow 11634.0) and we have a table in slide 28 of this presentation: https://cernbox.cern.ch/index.php/s/FfAZonfSest7kqQ

**Comment by clacaputo** [comment]
@cmsbuild please test

**Comment by clacaputo** [comment]
test parameters:

- enable_test = profiling

**Comment by clacaputo** [comment]
@cmsbuild please test

**Comment by CeliaFernandez** [comment]
I checked with very reduced statistics, 2000-2500 events.

**Comment by clacaputo** [comment]
Hi @CeliaFernandez , could you please check the size increase on wf `11834.21` with the statistics that you have stated here https://github.com/cms-sw/cmssw/pull/36926#issuecomment-1047814466?
Thanks

**Comment by CeliaFernandez** [comment]
Hi @clacaputo , I'm working on checking the size increase with the mentioned workflow. However, I have seen that when I run it I get several warning/error messages which I'm not pretty sure if they come from the implementation, the workflow itself or some bad configuration that I'm setting... For example, for step 3 (the one mentioned above):

`%MSG-w TrackNaN:  TrackProducer:pixelLessStepTracks 01-Mar-2022 18:34:17 CET  Run: 1 Event: 3`
`Track has NaN or the cov is not pos-definite`
`%MSG`
`%MSG-w BasicTrajectoryState:  MkFitOutputConverter:pixelLessStepTrackCandidates  01-Mar-2022 18:34:27 CET Run: 1 Event: 5`
`local error not pos-def`
`[  8.60678e-07 1.75447e-09 1.00676e-06 3.26687e-07-8.93862e-07`
`   1.75447e-09-1.29687e-08-1.59311e-07-3.21759e-08-1.45759e-07`
`   1.00676e-06-1.59311e-07 3.38469e-06 1.80004e-06 9.90766e-06`
`   3.26687e-07-3.21759e-08 1.80004e-06 4.32012e-06-2.41421e-05`
`  -8.93862e-07-1.45759e-07 9.90766e-06-2.41421e-05  0.00127064 ]`
`pos/mom/mf  (-28.4537,-19.378,-90.6192)   (-1.36899,-0.722564,-7.87008)   (0.00539667,0.00367532,3.7947)`

Is this normal?
Thanks!

**Comment by jpata** [comment]
this is a known issue https://github.com/cms-sw/cmssw/issues/35798 and should not affect your studies

**Comment by CeliaFernandez** [comment]
I compute the sizes again with the 11834.21 workflow. I did it with 1k events, which is the number used to compute the numbers above (I revisited the computation and I found that only 1k events were used), and also with a total of 16k which is the number of events we produced to obtain the validation plots. For the **1k sample** I obtain a **+1%** relative increase, and with the **16k** sample, a **+0.97%** relative increase. These numbers doesn't match with yours... Maybe we are missing something and we are underestimating the increase in size?  

We are computing these numbers with the methods provided in **printSizes.C** macro, that you can find here: https://root.cern/doc/v610/printSizes_8C_source.html
The measurement is done with respect to the tree stored inside the root file, not the root file itself...

I also attach 2 txt files with the space taken by each branch in case you want to check:
[info_1k.txt](https://github.com/cms-sw/cmssw/files/8199020/info_1k.txt)
[info_15974.txt](https://github.com/cms-sw/cmssw/files/8199021/info_15974.txt)

**Comment by clacaputo** [comment]
Hi @CeliaFernandez , thanks for having performed the measurement using `11834.21`. I'm inclined to consider your result more indicative, considering the statistics that you have used, while the one obtained by cmstest has less stat and therefore the compression is still not visible.
If you have the files somewhere, I can definitely repeat the measurement, although **printSizes.C** seems appropriate.

So, considering that the size increase is `~1` there is still open the point made by Joosep (https://github.com/cms-sw/cmssw/pull/36926#issuecomment-1046999194)

>1.2% bigger AOD means 1.2% less MC stats for CMS overall. There is obviously some physics decision here, with more analyses being possible, but all other analyses potentially suffering a bit. This may be a PC-level decision. The best we can do from the proponents&reco side is to ensure that the increase is as small as reasonably possible.

Did  @cms-sw/muon-pog-l2 have the chance to start a discussion with @exo ?

**Comment by jpata** [comment]
type muon

**Comment by jpata** [comment]
since you opened https://github.com/cms-sw/cmssw/pull/37805, can you close this one here?

**Pull request closed without merging**